### PR TITLE
Add first test

### PR DESCRIPTION
**Pull request by relud**
Will be left for a separate PR

- contributing docs
- maybe allow `expect.csv`
- use OrderedDict to compare results
- handle `expect.sql` for comparing results indirectly
- look for `expect.schema.*` or `{query_name}.schema.*` to define how to coerce non-json types and optionally pre-create destination table
- indicate cluster columns in schema field description
- indicate columns to omit from result in schema field description
- read `{query_name}/{test_name}/test.yaml` and `{query_name}/{test_name}.yaml` for additional test info.

### CODE REVIEW

File: tests/util.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            yield {
+                key: list(coerce_result(*value))
+                if isinstance(value, list)
+                else next(coerce_result(value))
+                for key, value in element.items()
+                # drop generated_time column
+                if key not in ("generated_time",)
+            }
+        elif isinstance(element, date):
+            yield element.strftime("%Y-%m-%d")
```
Comment (relud) [inline tests/util.py]: alternately we could require the `expect` to be in yaml format when it needs to support dates

File: tests/util.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+def coerce_result(*elements: Any):
+    for element in elements:
+        if isinstance(element, (dict, bigquery.Row)):
+            yield {
+                key: list(coerce_result(*value))
+                if isinstance(value, list)
+                else next(coerce_result(value))
+                for key, value in element.items()
+                # drop generated_time column
+                if key not in ("generated_time",)
```
Comment (relud) [inline tests/util.py]: `generated_time` is currently only validated by the destination table schema if it was loaded (so for this test it's required, a datetime, and in the correct column order). we could go further and assert that it's between the start and end of the test, and maybe the same for all output rows? I'm not sure if that's worth adding to this PR

File: tests/test_generated.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    except NotFound:
+        bq.create_dataset(dataset_id)
+    # wait for test
+    yield bq.dataset(dataset_id)
+    # clean up
+    try:
+        bq.delete_dataset(dataset_id).result()
+    except BadRequest as e:
+        if str(e).endswith("still in use."):
+            pass  # ignore still in use exception on delete_dataset
```
Comment (relud) [inline tests/test_generated.py]: iiuc this exception indicates that there are still tables in this dataset, so either `tables` needs to clean up, or there might be a setting to recursively delete

File: tests/test_generated.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from .util import coerce_result, generate_tests
+
+import json
+import os
+import pytest
+
+
+@pytest.fixture(scope="session")
+def bq():
+    return bigquery.Client(project=os.environ.get("GCP_PROJECT", None))
```
Comment (relud) [inline tests/test_generated.py]: there's probably a better named env var to use here

File: tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        "mode": "REQUIRED"
+      }
+    ]
+  },
+  {
+    "name": "client_id",
+    "type": "STRING",
+    "mode": "REQUIRED"
+  }
+]
```
Comment (relud) [inline tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json]: I made this support `yaml` and `ndjson` schemas as well, but `json` is supported by the `bq` cli tool, so I'm not sure if it's better to stick with `json` even though `yaml` is easier to maintain.

File: tests/clients_last_seen_v1/test_single_day/expect.ndjson [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{"submission_date":"2019-01-02","active_hours_sum":0.0,"attribution":{"source":"prev"},"client_id":"a","last_seen_date":"2019-01-01"}
+{"submission_date":"2019-01-02","active_hours_sum":1.0,"attribution":{"source":"test"},"client_id":"b","last_seen_date":"2019-01-02"}
+{"submission_date":"2019-01-02","active_hours_sum":1.0,"attribution":{"source":"test"},"client_id":"c","last_seen_date":"2019-01-02"}
```
Comment (relud) [inline tests/clients_last_seen_v1/test_single_day/expect.ndjson]: we can expect in `ndjson`, `json` or `yaml` format right now. I'm somewhat inclined to switch to `yaml` for date support, but the `bq load` can use this if we keep it as `ndjson` 🤷‍♀️

File: tests/util.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            try:
+                expect = load(resource_dir, "expect")
+            except FileNotFoundError:
+                continue
+
+            # generate tables for files with a supported table extension
+            for resource in next(os.walk(resource_dir))[2]:
+                if resource in ("expect.ndjson", "query_params.ndjson"):
+                    continue  # expect and query_params are not tables
+                table_name, extension = resource, ""
```
Comment (relud) [inline tests/util.py]: this should move above line 130, and line 130 should use `table_name` instead of `resource`

File: tests/util.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+@dataclass
+class GeneratedTest:
+    expect: List[Dict[str, Any]]
+    name: str
+    query: str
+    query_name: str
+    query_params: List[Any]
+    replace: Dict[str, str]
+    tables: Set[Table]
```
Comment (relud) [inline tests/util.py]: using `Set[Table]` isn't really better than `List[Table]` or perhaps `Dict[str, Table]`, and requires table to be frozen, which makes Table.get_schema more complicated than necessary. Should probably switch to `list` or `dict`.

File: tests/util.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    def get_schema(self):
+        resource_dir, resource = os.path.split(self.source_path)
+        full_name, _ = resource.rsplit(".", 1)
+        try:
+            return [
+                bigquery.SchemaField.from_api_repr(field)
+                for field in load(resource_dir, f"{full_name}.schema")
+            ]
+        except FileNotFoundError:
+            return None
```
Comment (relud) [inline tests/util.py]: should unfreeze `Table` and set `Table.schema` in a `post_init` function

File: tests/test_generated.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                        field=field.name
+                    )
+                    break  # stop because there can only be one time partitioning field
+        with open(table.source_path, "rb") as file_obj:
+            job = bq.load_table_from_file(file_obj, destination, job_config=job_config)
+        job.result()
+    # clean up handled by default_dataset fixture
+
+
+def test_generated(bq, dataset, generated_test):
```
Comment (relud) [inline tests/test_generated.py]: this is complicated enough that I think a `tests/README.md` is necessary and a good place to indicate how generated tests are laid out

File: tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
+[
+  {
+    "name": "submission_date_s3",
+    "type": "DATE",
+    "mode": "REQUIRED",
+    "description": "time_partitioning_field"
```
Comment (relud) [inline tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json]: RFC on using this to locate `time_partitioning_field` and perhaps `clustering_field`s or moving it to a separate file, which could be merged with `query_params.yaml` into a `test_info.yaml` or something? that would prevent the proliferation of small files for additional test info and in small enough cases could be an appropriate place for all of the data for a test.

File: tests/test_generated.py (hunk 4 above)
Comment (jklukas) [reply tests/test_generated.py]: Is there no python equivalent to the [Java SDK's testing helpers](https://github.com/googleapis/google-cloud-java/blob/master/TESTING.md#testing-code-that-uses-bigquery)? With the Java machinery, it's able to read creds from env vars and you can then extract the default project id based on the creds.

File: tests/test_generated.py (hunk 3 above)
Comment (jklukas) [reply tests/test_generated.py]: Recursively deleting a dataset is something the Java helpers handle, so I'm guessing not built straight into the API.

File: tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json (hunk 11 above)
Comment (jklukas) [reply tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json]: This is indeed looking like quite a bit of configuration to manage for writing a test, so making some optimizations for easier test writing is appropriate. Putting time_partitioning_field in description seems reasonable and self-documenting.

File: tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json (hunk 5 above)
Comment (jklukas) [reply tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json]: My vote is to stick with json for this due to it being an existing standard that interoperates with `bq`.

File: tests/clients_last_seen_v1/test_single_day/expect.ndjson (hunk 6 above)
Comment (jklukas) [reply tests/clients_last_seen_v1/test_single_day/expect.ndjson]: I vote for sticking with what `bq` supports here, too. Coercing between dates and strings is simple enough and I don't anticipate other type compatibility problems.

File: tests/util.py (hunk 2 above)
Comment (jklukas) [reply tests/util.py]: I'm fine cutting this out of scope for this PR.

File: tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json (hunk 11 above)
Comment (relud) [reply tests/clients_last_seen_v1/test_single_day/clients_daily_v6.schema.json]: okay, i'll leave it here for now, and in a follow-up PR add support for reading `tests/clients_last_seen_v1/test_*.yaml` and `tests/clients_last_seen_v1/test_*/test.yaml` as a place to put things, which should also enable `time_partitioning_field` with `autodetect`

File: tests/test_generated.py (hunk 4 above)
Comment (relud) [reply tests/test_generated.py]: Due to the priority of various things there's no built in env var that can override the gcloud cli default project but still use the gcloud default creds. default creds and not setting `GCP_PROJECT` also works fine. I'll extract this to a later PR for now.

File: tests/test_generated.py (hunk 3 above)
Comment (relud) [reply tests/test_generated.py]: I just needed to set `delete_contents=True`

File: tests/util.py (hunk 2 above)
Comment (relud) [reply tests/util.py]: I'll make sure to thoroughly explain it in `tests/README.md`. I might add a kwarg later to control this, or do something with schema field descriptions indicating what to drop for expect.

File: tests/util.py (hunk 1 above)
Comment (relud) [reply tests/util.py]: I'm going to file a follow-up PR to load `expect.schema` or `{query_name}.schema` to cast types not available in JSON.

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    - run:
+        name: PyTest with linters
+        # Google's client libraries will check for GOOGLE_APPLICATION_CREDENTIALS
+        # and use a file in that location for credentials if present;
+        # See https://cloud.google.com/docs/authentication/production
+        command: |
+          echo "${GCLOUD_SERVICE_KEY:?}" > "$GOOGLE_APPLICATION_CREDENTIALS"
+          venv/bin/pytest --black --docstyle --flake8 --mypy-ignore-missing-imports
+        environment:
+          GOOGLE_APPLICATION_CREDENTIALS: /tmp/gcp.json
```
Comment (jklukas) [inline .circleci/config.yml]: Can you move the environment section above the command section?

File: tests/README.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  - File extensions `yaml`, `json` and `ndjson` are supported
+  - Formats other than `.json` should not be used because they are not
+    supported by `bq load`
+- Query parameters
+  - Scalar query params should be defined as a dict with keys `name`, `type` or
+    `type_`, and `value`
+    - Query params with other keys will be passed to `.from_api_repr` for
+      each available type and must be valid for one of them
+  - `query_parameters.yaml` may be used instead of `query_params.yaml`, but
+    they are mutually exclusive
```
Comment (jklukas) [inline tests/README.md]: Why support both? `query_params.yaml` seems fine to me.

Comment (relud) [reply tests/README.md]: okay, removed support for the alternative

**Pull request merged**